### PR TITLE
610 implement initialization scheme for flows

### LINKED ISSUE #610: Implement initialization scheme for flows
To run a simulation using the IDE model we need the flows describing the transitions from one compartment to another one for a sufficient amount of time points before the start of the simulation. We can access data that describes the flow from InfectedNoSymptoms to InfectedSymptoms and we want to compute the remaining flows with this given data. This is a two-step scheme that is described in detail in the corresponding Overleaf Document.

### PR DESCRIPTION
**Pull request by lenaploetzke**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

- Added functionality to calculate initial flows based on real data for an IDE-SECIR model.
- Added functionality to get mean values to StateAgeFunction.
- Added tests and an example.

If need be, add additional information and what the reviewer should look out for in particular:
-/-
## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [ ] Corresponding issue(s) is/are linked and addressed
- [ ] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [ ] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [ ] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [ ] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

Closes #610.

### CODE REVIEW

File: cpp/examples/ide_initialization.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    // Check provided parameters.
+    std::string filename = setup(argc, argv);
+    if (filename.empty()) {
+        std::cout << "You did not provide a valid filename. A default initialization is used." << std::endl;
+
+        using Vec = mio::TimeSeries<ScalarType>::Vector;
+        mio::TimeSeries<ScalarType> init((int)mio::isecir::InfectionTransition::Count);
+        init.add_time_point<Eigen::VectorXd>(-7., Vec::Constant((int)mio::isecir::InfectionTransition::Count, 1.));
+        while (init.get_last_time() < 0) {
+            init.add_time_point(init.get_last_time() + dt,
```
Comment (annawendler) [inline cpp/examples/ide_initialization.cpp:78]: Have you checked that the example doesn't lead weird behaviour as the other example used to? Should a factor dt be added here as well?

File: cpp/examples/ide_secir.cpp [hunk 2]
```diff
@@ -72,11 +72,13 @@ int main()
     mio::SmootherCosine smoothcos(2.0);
     mio::StateAgeFunctionWrapper delaydistribution(smoothcos);
     std::vector<mio::StateAgeFunctionWrapper> vec_delaydistrib(num_transitions, delaydistribution);
-    vec_delaydistrib[(int)mio::isecir::InfectionTransition::SusceptibleToExposed].set_parameter(3.0);
+    // TransitionDistribution for SusceptibleToExposed is not used.
+    vec_delaydistrib[(int)mio::isecir::InfectionTransition::SusceptibleToExposed].set_parameter(-1.);
```
Comment (annawendler) [inline cpp/examples/ide_secir.cpp:76]: I would either not set this parameter explicitly or comment that this can be set arbitrarily.

File: cpp/memilio/epidemiology/state_age_function.h [hunk 3]
```diff
@@ -58,6 +58,12 @@ namespace mio
  * decreasing. This is no limitation as the support is only needed for StateAgeFunctions of Type a) as given above.
  * For classes of type b) a dummy implementation logging an error and returning -2 for get_support_max() should be implemented.
  *
+ * The get_mean method is virtual and implements a basic version to determine the mean value of the StateAgeFunction. 
+ * The base class implementation uses the fact that the StateAgeFunction is a survival function
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:62]: Not every StateAgeFunction is a survival function, only StateAgeFunctions of type a). If the get_mean method relies on the property that StateAgeFunction is a survival function and can only be used for StateAgeFunctions of type a), this should be mentioned here.

File: cpp/memilio/epidemiology/state_age_function.h [hunk 4]
```diff
@@ -69,6 +75,8 @@ struct StateAgeFunction {
      */
     StateAgeFunction(ScalarType init_parameter)
         : m_parameter{init_parameter}
+        , m_mean{-1.} // Initialize mean as not set.
+        , m_mean_tol{-1.} // Initialize mean as not set.
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:79]:
```suggestion
        , m_mean_tol{-1.} // Initialize tolerance for computation of mean as not set.
```

File: cpp/memilio/epidemiology/state_age_function.h [hunk 5]
```diff
@@ -69,6 +75,8 @@ struct StateAgeFunction {
...
         , m_support_max{-1.} // initialize support maximum as not set
         , m_support_tol{-1.} // initialize support tolerance as not set
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:81]:
```suggestion
        , m_support_max{-1.} // Initialize support maximum as not set.
        , m_support_tol{-1.} // Initialize tolerance for computation of support as not set.
```
For consistency

File: cpp/memilio/epidemiology/state_age_function.h [hunk 6]
```diff
@@ -176,6 +185,37 @@ struct StateAgeFunction {
...
 
+    /**
+     * @brief Computes the mean value of the function using the time step size dt and some tolerance tol. 
+     * 
+     * This is a basic version to determine the mean value of a survival function
+     * through numerical integration of the integral that describes the expected value.
+     *
+     * For some specific derivations of StateAgeFunction%s there are more efficient ways to determine the 
+     * the mean value which is why this member function is virtual and can be overridden (see, e.g., ExponentialDecay).
+     * The mean value is only needed for StateAgeFunction%s that are used as TransitionDistribution%s.
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:196]: Can you make a reference to type a) here?

File: cpp/memilio/epidemiology/state_age_function.h [hunk 7]
```diff
@@ -176,6 +185,37 @@ struct StateAgeFunction {
...
+    virtual ScalarType get_mean(ScalarType dt = 1., ScalarType tol = 1e-10)
+    {
+        if (!floating_point_equal(m_mean_tol, tol, 1e-14) || floating_point_equal(m_mean, -1., 1e-14)) {
+            ScalarType mean         = 0.;
+            ScalarType supp_max_idx = std::ceil(get_support_max(dt, tol) / dt);
+            // Integration using Trapezoidal rule.
+            for (int i = 0; i < supp_max_idx; i++) {
+                mean += dt * eval(i * dt);
+            }
+            mean -= 0.5 * dt * eval(0 * dt);
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:211]: Why not like this:
`ScalarType mean = 0.5 * dt * eval(0 * dt);`
and then start the for loop at i=1? For me, this would be more intuitive.

File: cpp/memilio/epidemiology/state_age_function.h [hunk 8]
```diff
@@ -298,6 +356,10 @@ struct SmootherCosine : public StateAgeFunction {
         return m_support_max;
     }
 
+    // TODO: There is also a closed form for the mean value of Smoothercosine: 0.5*m_parameter.
+    // However, an StateAgeFunction that uses the default implementation is required for testing purposes.
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:360]:
```suggestion
    // However, a StateAgeFunction that uses the default implementation is required for testing purposes.
```

File: cpp/memilio/epidemiology/state_age_function.h [hunk 9]
```diff
@@ -365,6 +427,22 @@ struct ConstantFunction : public StateAgeFunction {
...
+    /**
+     * @brief Computes the mean value of the function. 
+     * 
+     * For ConstantFunction, the mean value is the function parameter.
+     *
+     * @param[in] dt Time step size used for the numerical integration (unused for ConstantFunction). 
+     * @param[in] tol The maximum support used for numerical integration is calculated using this tolerance (unused for ConstantFunction). 
+     * @return ScalarType mean value.
+     */
+    ScalarType get_mean(ScalarType dt = 1., ScalarType tol = 1e-10) override
```
Comment (annawendler) [inline cpp/memilio/epidemiology/state_age_function.h:439]: I don't think we should implement a get_mean method for a StateAgeFunction of type b). 
At the moment, the get_mean() method is only called for StateAgeFunctions of type a) that are used as TransitonDistributions. If we call the get_mean method for a SAF of type b) something went wrong (at least for the current state of the IDE model). This is why I would argue to not implement a get_mean method here and log an error, even if it is really easy to compute the mean.

File: cpp/models/ide_secir/model.cpp [hunk 10]
```diff
@@ -174,20 +179,17 @@ void Model::compute_flow(int idx_InfectionTransitions, Eigen::Index idx_Incoming
     Eigen::Index calc_time_index = (Eigen::Index)std::ceil(
         parameters.get<TransitionDistributions>()[idx_InfectionTransitions].get_support_max(dt, m_tol) / dt);
 
-    Eigen::Index num_time_points = m_transitions.get_num_time_points();
+    for (Eigen::Index i = current_time_index - calc_time_index; i < current_time_index; i++) {
+        // (current_time_index - i) * dt is the time, the individuals has already spent in this state.
```
Comment (annawendler) [inline cpp/models/ide_secir/model.cpp:183]:
```suggestion
        // (current_time_index - i) * dt is the time the individuals have already spent in this state.
```

File: cpp/models/ide_secir/model.h [hunk 11]
```diff
@@ -59,37 +59,41 @@ class Model
...
     {
         if (!((int)m_transitions.get_num_elements() == (int)InfectionTransition::Count)) {
-            log_error("A variable given for model construction is not valid. Number of elements in transition vector "
-                      "does not match the required number.");
+            log_error(
+                "Initialization failed. Number of elements in transition vector does not match the required number.");
+            return true;
+        }
+
+        // We have some things that just work out if the last time point is zero, e.g. the first time point in m_populations is set to zero ion the constructor.
```
Comment (annawendler) [inline cpp/models/ide_secir/model.h:67]:
```suggestion
        // We have some things that just work out if the last time point is zero, e.g. the first time point in m_populations is set to zero in the constructor.
```
With respect to PR #979 this can already be removed here in my opinion.

File: cpp/models/ide_secir/model.h [hunk 12]
```diff
@@ -118,18 +122,30 @@ class Model
     void compute_susceptibles(ScalarType dt);
 
     /**
-     * @brief Computes size of a flow.
+     * @brief Computes size of a flow for the current last time value in m_transitions.
      * 
      * Computes size of one flow from #InfectionTransition, specified in idx_InfectionTransitions, for the current 
-     * last time value in m_transitions.
+     * last time value in m_transitions. 
+     * The function compute_flow with one parameter more is called with the specification of current_time_index as
```
Comment (annawendler) [inline cpp/models/ide_secir/model.h:129]:
```suggestion
     * The function compute_flow with an additional parameter is called with the specification of current_time_index as 
```

File: cpp/models/ide_secir/model.h [hunk 13]
```diff
@@ -118,18 +122,30 @@ class Model
...
     /**
-     * @brief Computes size of a flow.
+     * @brief Computes size of a flow for the current last time value in m_transitions.
      * 
      * Computes size of one flow from #InfectionTransition, specified in idx_InfectionTransitions, for the current 
-     * last time value in m_transitions.
+     * last time value in m_transitions. 
+     * The function compute_flow with one parameter more is called with the specification of current_time_index as 
+     * the last index of m_transitions.
+     * See also compute_flow for a description of the parameters.
```
Comment (annawendler) [inline cpp/models/ide_secir/model.h:131]: I would change the order of the compute_flow functions. I think it is easier to understand that way.

File: cpp/models/ide_secir/model.h [hunk 14]
```diff
@@ -118,18 +122,30 @@ class Model
...
+     */
+    void compute_flow(Eigen::Index idx_InfectionTransitions, Eigen::Index idx_IncomingFlow, ScalarType dt);
+
+    /**
+     * @brief Computes size of a flow.
+     * 
+     * Computes size of one flow from #InfectionTransition, specified in idx_InfectionTransitions, for the ctime index current_time_index.
```
Comment (annawendler) [inline cpp/models/ide_secir/model.h:138]:
```suggestion
     * Computes size of one flow from #InfectionTransition, specified in idx_InfectionTransitions, for the time index current_time_index.
```

File: cpp/models/ide_secir/model.h [hunk 15]
```diff
@@ -197,6 +214,18 @@ class Model
      */
     void compute_recovered();
 
+    /**
+     * @brief Getter for the global support_max, i.e. the maximum of support_max over all TransitionDistributions.
+     *
+     * This determines how many inital values we need for the flows.
```
Comment (annawendler) [inline cpp/models/ide_secir/model.h:220]: Here it would make sense to say that we don't necessarily need that many flows but then it is ensured that we have enough initial values (as you did in the check_constraints function). Or I would state it differently and say that this determines how many initial flows we compute.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+
+#include <string>
+#include <cmath>
+
+namespace mio
+{
+namespace isecir
+{
+
+// We assume that we start the simulation at time 0 and want to compute the necessary flows
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:41]: I think it would be fairly easy to adapt this, but it is not a problem at all to leave this assumption.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+    ScalarType global_support_max         = model.get_global_support_max(dt);
+    Eigen::Index global_support_max_index = Eigen::Index(std::ceil(global_support_max / dt));
+
+    // m_transitions should be empty at the beginning.
+    if (model.m_transitions.get_num_time_points() > 0) {
+        model.m_transitions = TimeSeries<ScalarType>((int)InfectionTransition::Count);
+    }
+
+    // The first time we need is -4 * global_support_max.
+    Eigen::Index start_shift = -4 * global_support_max_index;
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:72]: I would find it more intuitive to define start_shift as a positive value. This would also be consistent with the other index_shift somewhere below which is also a positive value.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+    // The last time needed is dependent on the mean stay times in the Exposed compartment and in InfectedNoSymptoms
+    // before transition to InfectedSymptoms.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:74]: What do you mean with this half-sentence?

File: cpp/models/ide_secir/parameters_io.cpp [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+    Eigen::Index last_time_index_needed = Eigen::Index(std::ceil(
+        (model.parameters.get<TransitionDistributions>()[Eigen::Index(InfectionTransition::ExposedToInfectedNoSymptoms)]
+             .get_mean(dt) +
+         model.parameters
+             .get<TransitionDistributions>()[Eigen::Index(InfectionTransition::InfectedNoSymptomsToInfectedSymptoms)]
+             .get_mean(dt)) /
+        dt));
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:81]: Below you are defining `mean_InfectedNoSymptomsToInfectedSymptoms`. You could already do this here, I think this would make the connection clearer.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+    for (Eigen::Index i = start_shift; i <= last_time_index_needed; i++) {
+        // Add time point.
+        model.m_transitions.add_time_point(
+            i * dt, TimeSeries<ScalarType>::Vector::Constant((int)InfectionTransition::Count, 0.));
+    }
+
+    //--- Calculate the flow InfectedNoSymptomsToInfectedSymptoms using the RKI data and store in the m_transitions object.---
+    ScalarType min_offset_needed = std::ceil(
+        model.m_transitions.get_time(0) -
+        1); //Need -1 if first time point is integer and just the floor value if not, therefore use ceil and -1
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:92]:
```suggestion
        1); //Need -1 if first time point is integer and just the ceil value if not, therefore use ceil and -1
```
Is this what you mean?

File: cpp/models/ide_secir/parameters_io.cpp [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+    ScalarType max_offset_needed = std::ceil(model.m_transitions.get_last_time());
+
+    bool min_offset_needed_avail = false;
+    bool max_offset_needed_avail = false;
+    // Go through the entries of rki_data and check if date is needed for calculation. Confirmed cases are scaled.
+    // Define dumy variables to store the first and the last index of the TimeSeries where the considered entry of rki_data is potentially needed.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:98]:
```suggestion
    // Define dummy variables to store the first and the last index of the TimeSeries where the considered entry of rki_data is potentially needed.
```

File: cpp/models/ide_secir/parameters_io.cpp [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+    ScalarType time_idx           = 0;
+    for (auto&& entry : rki_data) {
+        int offset = get_offset_in_days(entry.date, date);
+        if (!(offset < min_offset_needed) && !(offset > max_offset_needed)) {
+            if (offset == min_offset_needed) {
+                min_offset_needed_avail = true;
+            }
+            // Smallest index for which the entry is needed is when offset=std::ceil(time_idx).
+            idx_needed_first =
+                Eigen::Index(std::max(std::floor((offset - model.m_transitions.get_time(0) - 1) / dt), 0.));
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:110]: Here, comment and definition of `idx_needed_first` do not match. Same for idx_needed_last.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+            // Biggest index for which the entry is needed is when offset=std::floor(time_idx-dt).
+            idx_needed_last = Eigen::Index(std::min(std::ceil((offset - model.m_transitions.get_time(0) + 1) / dt),
+                                                    double(model.m_transitions.get_num_time_points() - 1)));
+
+            for (Eigen::Index i = idx_needed_first; i <= idx_needed_last; i++) {
+                time_idx = model.m_transitions.get_time(i);
+                if (offset == int(std::floor(time_idx))) {
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:117]: For the interpolation of the data: This is the same scheme as in the LCT model right? I think it would make sense to define a separate function and use it both in the LCT and IDE model. Then it could easily be used for the ODE model as well and code duplication would be avoided.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+        // Compute flow InfectedCriticalToRecovered.
+        model.compute_flow(Eigen::Index(InfectionTransition::InfectedCriticalToRecovered),
+                           Eigen::Index(InfectionTransition::InfectedSevereToInfectedCritical), dt, i - start_shift);
+        // Compute flow InfectedCriticalToDead.
+        model.compute_flow(Eigen::Index(InfectionTransition::InfectedCriticalToDead),
+                           Eigen::Index(InfectionTransition::InfectedSevereToInfectedCritical), dt, i - start_shift);
+    }
+
+    //--- Calculate the remaining flows. ---
+    // Compute flow InfectedNoSymptomsToInfectedSymptoms for -global_support_max, ..., 0.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:181]:
```suggestion
    // Compute flow ExposedToInfectedNoSymptoms for -global_support_max, ..., 0.
```

File: cpp/models/ide_secir/parameters_io.cpp [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+    // Use mean value of the TransitionDistribution InfectedNoSymptomsToInfectedSymptoms for the calculation.
+    ScalarType mean_InfectedNoSymptomsToInfectedSymptoms =
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:183]: You could already define this earlier, see above.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+    Eigen::Index index_shift_mean = Eigen::Index(std::round(mean_InfectedNoSymptomsToInfectedSymptoms / dt));
+    for (Eigen::Index i = -global_support_max_index; i <= 0; i++) {
+        model.m_transitions[i - start_shift][Eigen::Index(InfectionTransition::ExposedToInfectedNoSymptoms)] =
+            (1 / model.parameters.get<TransitionProbabilities>()[Eigen::Index(
+                     InfectionTransition::InfectedNoSymptomsToInfectedSymptoms)]) *
+            model.m_transitions[i - start_shift + index_shift_mean]
+                               [Eigen::Index(InfectionTransition::InfectedNoSymptomsToInfectedSymptoms)];
+    }
+
+    // Compute flow ExposedToInfectedNoSymptoms for -global_support_max, ..., 0.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:196]:
```suggestion
    // Compute flow SusceptibleToExposed for -global_support_max, ..., 0.
```

File: cpp/models/ide_secir/parameters_io.cpp [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+    // Use mean values of the TransitionDistribution ExposedToInfectedNoSymptoms and of the TransitionDistribution InfectedNoSymptomsToInfectedSymptoms for the calculation.
+    ScalarType mean_ExposedToInfectedNoSymptoms =
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:198]: Same as above.

File: cpp/models/ide_secir/parameters_io.h [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+*
+* The flows InfectedSymptomsToInfectedSevere, InfectedSymptomsToRecovered, InfectedSevereToInfectedCritical,
+* InfectedSevereToRecovered, InfectedCriticalToDead and InfectedCriticalToRecovered can then be calculated using
+* the InfectedNoSymptomsToInfectedSymptoms flow with the standard formula from the IDE model.
+* The ExposedToInfectedNoSymptoms and InfectedNoSymptomsToInfectedSymptoms flows are calculated 
+* using the means of the respective TransitionDistribution. 
+* The flow InfectedNoSymptomsToInfectedSymptoms is calculated with the standard formula from the IDE model
+* using the results for ExposedToInfectedNoSymptoms.
+*
+* The number of death used in the model is set to the number given in the RKI data.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.h:55]:
```suggestion
* The number of deaths used in the model is set to the number given in the RKI data.
```

File: cpp/models/ide_secir/parameters_io.h [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+* We also set the number of confirmed cases in the model.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.h:56]:
```suggestion
* We also set the number of total confirmed cases in the model. 
```

File: cpp/tests/test_ide_parameters_io.cpp [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+TEST(TestIDEParametersIo, RKIcompareWithPreviousRun)
+{
+    // Define start date and the total population used for the initialization.
+    ScalarType total_population = 80 * 1e6;
+    auto start_date             = mio::Date(2020, 6, 3);
+    ScalarType dt               = 0.5;
+    // Initialize model.
+    mio::isecir::Model model(mio::TimeSeries<ScalarType>(0., mio::TimeSeries<ScalarType>::Vector::Constant(
+                                                                 (int)mio::isecir::InfectionTransition::Count, 1.)),
+                             total_population, -1);
```
Comment (annawendler) [inline cpp/tests/test_ide_parameters_io.cpp:44]: Why do you set total_confirmed_cases to -1?

File: cpp/tests/test_ide_parameters_io.cpp [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+// Check some cases where computation of initial values for an IDE model based on RKI data should fail.
+TEST(TestIDEParametersIo, ParametersIoRKIFailure)
+{
+    // Define start date and the total population used for the initialization.
+    ScalarType total_population = 80 * 1e6;
+    ScalarType dt               = 0.5;
+
+    // Initialize model.
+    mio::isecir::Model model(mio::TimeSeries<ScalarType>((int)mio::isecir::InfectionTransition::Count),
+                             total_population, -1);
```
Comment (annawendler) [inline cpp/tests/test_ide_parameters_io.cpp:97]: As above: why total_confirmed_cases=-1?

File: cpp/tests/test_ide_secir.cpp [hunk 32]
```diff
@@ -399,17 +398,41 @@ TEST(IdeSecir, testModelConstraints)
     auto constraint_check = model_wrong_size.check_constraints(dt);
     EXPECT_TRUE(constraint_check);
 
-    // --- Test with too few time points.
+    // --- Test if the last time point is not zero.
```
Comment (annawendler) [inline cpp/tests/test_ide_secir.cpp:401]: With respect to PR #979 I would remove this test already here.

File: cpp/tests/test_state_age_function.cpp [hunk 33]
```diff
@@ -127,24 +121,47 @@ TEST(TestStateAgeFunction, testGetSupportMax)
...
-    mio::StateAgeFunctionWrapper wrapper(smoothcos);
+    EXPECT_NEAR(smoothcos.get_mean(dt), 0.5, 1e-14);
+    smoothcos.set_parameter(2.0);
+    EXPECT_NEAR(smoothcos.get_mean(dt), 1.0, 1e-14);
+}
+
+TEST(TestStateAgeFunction, testSAFWrapperSpecialMember)
+{
+    ScalarType dt = 0.5;
+    mio::ExponentialDecay expdecay(1.0);
```
Comment (annawendler) [inline cpp/tests/test_state_age_function.cpp:152]: Why did you change the type of the StateAgeFunction here?

File: cpp/tests/test_state_age_function.cpp [hunk 34]
```diff
@@ -204,16 +229,18 @@ TEST(TestStateAgeFunction, testSAFWrapperSettersAndGetters)
 
     EXPECT_EQ(wrapper.get_parameter(), testvalue_before);
     EXPECT_EQ(wrapper.get_support_max(dt), testvalue_before);
+    EXPECT_EQ(wrapper.get_mean(dt), testvalue_before * 0.5);
```
Comment (annawendler) [inline cpp/tests/test_state_age_function.cpp:232]: A comment that this is the formula for the mean for SmootherCosine would be helpful here (although this is mentioned in the comment for get_mean in SmootherCosine).

**Review by annawendler — COMMENTED** [review]
Looks very good!
I made some comments and suggestions.

File: cpp/examples/ide_initialization.cpp (hunk 1 above)
Comment (lenaploetzke) [reply cpp/examples/ide_initialization.cpp:78]: Yes, this is not really a problem here but i added the factor

File: cpp/models/ide_secir/parameters_io.cpp (hunk 18 above)
Comment (lenaploetzke) [reply cpp/models/ide_secir/parameters_io.cpp:72]: I wanted to make clear that the mean stay time of individuals in InfectedNoSymptoms is needed for pre-symptomatic individuals and the people who recover directly are not relevant (so the mean of the transition InfectedNoSymptomsToInfectedSymptoms). I have rewritten it a bit, is it better now?

File: cpp/models/ide_secir/parameters_io.cpp (hunk 19 above)
Comment (lenaploetzke) [reply cpp/models/ide_secir/parameters_io.cpp:79]: Good idea

File: cpp/models/ide_secir/parameters_io.cpp (hunk 20 above)
Comment (lenaploetzke) [reply cpp/models/ide_secir/parameters_io.cpp:90]: Actually, floor is right here. If model.m_transitions.get_time(0) is not an integer, i need floor(model.m_transitions.get_time(0)) for the calculation which is equal to ceil(model.m_transitions.get_time(0)-1).
If model.m_transitions.get_time(0) is an integer, i need model.m_transitions.get_time(0)-1 for the calculation.

File: cpp/models/ide_secir/parameters_io.cpp (hunk 23 above)
Comment (lenaploetzke) [reply cpp/models/ide_secir/parameters_io.cpp:115]: The concept for calculating a value for num_confirmed at a non-integer time point is the same (simple linear interpolation). However, the values are processed differently (compartments vs flows that are calculated) and the calculation methods do not match. Since we have to go through the entries in a while loop, it also makes no sense to put the linear interpolation in a function, so I see no code duplication here or no way to generalize the concepts.

File: cpp/tests/test_ide_parameters_io.cpp (hunk 30 above)
Comment (lenaploetzke) [reply cpp/tests/test_ide_parameters_io.cpp:44]: Actually, i should just use the default value for this.

**Comment by annawendler** [comment]
Note: Once this PR is merged the links will work in PR #1000 and it can be merged.

File: cpp/models/ide_secir/parameters_io.cpp [hunk 35]
```diff
@@ -0,0 +1,229 @@
...
+namespace mio
+{
+namespace isecir
+{
+
+IOResult<void> set_initial_flows(Model& model, ScalarType dt, std::string const& path, Date date,
+                                 ScalarType scale_confirmed_cases)
+{
+    //--- Preparations ---
+    // Try to get rki data from path.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.cpp:45]:
```suggestion
    // Try to get RKI data from path.
```

File: cpp/models/ide_secir/parameters_io.h [hunk 36]
```diff
@@ -0,0 +1,76 @@
...
+* The flow InfectedNoSymptomsToInfectedSymptoms is calculated with the standard formula from the IDE model
+* using the results for ExposedToInfectedNoSymptoms.
+*
+* The number of deaths used in the model is set to the number given in the RKI data.
+* We also set the number of total confirmed cases in the model. 
+* Therefore the initialization method using the total confirmed cases is used in the model. See also the documentation of the model class.
+* 
+* The start date of the model simulation is set to t0=0.
+*
+* @param[in, out] model The model for which the inital flows should be computed.
```
Comment (annawendler) [inline cpp/models/ide_secir/parameters_io.h:61]:
```suggestion
* @param[in, out] model The model for which the initial flows should be computed.
```

File: cpp/tests/test_state_age_function.cpp [hunk 37]
```diff
@@ -204,16 +229,20 @@ TEST(TestStateAgeFunction, testSAFWrapperSettersAndGetters)
 
     EXPECT_EQ(wrapper.get_parameter(), testvalue_before);
     EXPECT_EQ(wrapper.get_support_max(dt), testvalue_before);
+    // Use the formula for the mean of a SmootherCosine StateAgeFunction: 0.5 * parameter.
+    // See also the Dokumentation of get_mean().
```
Comment (annawendler) [inline cpp/tests/test_state_age_function.cpp:233]:
```suggestion
    // See also the documentation of get_mean().
```

**Review by annawendler — CHANGES_REQUESTED** [review]

File: cpp/models/ide_secir/parameters_io.cpp (hunk 23 above)
Comment (lenaploetzke) [reply cpp/models/ide_secir/parameters_io.cpp:122]: Okay lets move this to a new issue.

**Review by annawendler — APPROVED** [review]
Looks good!
